keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1046 @@
+<<[/UNRESOLVED]>>
+
+During Alpha, this capability is behind the PodIdentityX509Support feature gate.
+
+### PodCertificate Projected Volume Sources
+
+The PodCertificate projected volume source instructs Kubelet to inject a
+credential bundle consisting of a private key and certificate chain into the
+container fileystem.  Kubelet handles generating the key, requesting a
+certificate from the named signer, and renewing the certificate before it
+expires.
+
+Kubelet supports only a few key generation strategies, each named by a string
+that encapsulates both the key type and any parameters necessary (for example,
+RSA modulus size).  The intention is to offer a a tasting menu of reasonable key
+choices, rather than offering a flexibility to a wide variety of parameters.
+Key types supported by Kubelet are "RSA2048", "RSA3072", "RSA4096", "ECDSAP256",
+and "ECDSAP384".  If no key type is specified, Kubelet defaults to "ECDSAP256".
+
+<<[UNRESOLVED]>>
+Should RSA2048 be offered?  I'm inclined to say no.  Are there any other key types that should be offered?
+<<[/UNRESOLVED]>>
+
+Once the key is generated, Kubelet creates a CertificateSigningRequest that
+contains an X.509 CertificateRequest with no data in the Name or
+SubjectAlternateName fields, but instead embeds an X.509 PodIdentity extension
+with the details of the pod that mounted the volume.  The signer is intended to treat this CertificateRequest merely as an indication that it should issue a certificate for the named pod, rather than trying to issue a certificate that follows the structure of the CertificateRequest.
+
+Kubelet then starts a single-object informer on the CertificateSigningRequest object it just created, and waits until the CSR is Approved and Issued.  It then takes the private key it generated and the issued certificate chain, and writes them to the pod filesystem in a single PEM file.  The first block in the PEM file contains the private key, and the remaining blocks contain the certificate chain in the same order they were in the CSR.
+
+Kubelet strips any inter-block comments and intra-block headers that were
+present in the issued certificate chain from the signer.
+
+Kubelet always sets a 24-hour TTL on its certificate requests, although of course the signer is free to ignore the requested TTL and return a certificate with any TTL.
+
+During the initial volume mount, before the pod has started, any errors returned during the certificate issuance process will block pod startup.
+
+Kubelet will begin trying to rotate the certificate once less than 50% of its TTL is remaining, or after 24 hours, whichever is sooner.  Errors in the certificate issuance process are treated as non-blocking while the original certificate is still valid.  Once the original certificate has expired, errors are treated as blocking, because it's better for the pod to noisily go unhealthy than to continue to run with an expired certificate.
```

Comment:
What does it mean that errors are "blocking," given that rotation is async? Does that mean we kill the Pod?

Reply:
```suggestion
The kubelet will begin trying to rotate the certificate once less than 50% of its TTL is remaining, or after 24 hours, whichever is sooner.  Errors in the certificate issuance process are treated as non-blocking while the original certificate is still valid.  Once the original certificate has expired, errors are treated as blocking, because it's better for the pod to noisily go unhealthy than to continue to run with an expired certificate.
```

Reply:
I just meant that we would start returning an error from the projected volume mount setup function, which is being polled every few seconds.  I don't know exactly what failure mode that will cause for the pod, but I believe it will be noisy at the very least (probably you'll get events emitted).

Reply:
Clarified to remove the "blocking/non-blocking" language.  Instead, I talk about whether or not errors are returned from the volume setup function.